src/entities/swapRefactorDraft.ts

```diff
@@ -0,0 +1,376 @@
+            );
+
+            amount =
+                this.swapKind === SwapKind.GivenIn
+                    ? TokenAmount.fromRawAmount(
+                          this.outputAmount.token,
+                          abs(
+                              result[
+                                  this.assets.indexOf(
+                                      this.convertNativeAddressToZero(
+                                          this.outputAmount.token.address,
+                                      ),
+                                  )
+                              ],
+                          ),
+                      )
+                    : TokenAmount.fromRawAmount(
+                          this.inputAmount.token,
+                          abs(
+                              result[
+                                  this.assets.indexOf(
+                                      this.convertNativeAddressToZero(
+                                          this.inputAmount.token.address,
+                                      ),
+                                  )
+                              ],
+                          ),
+                      );
+        } else {
+            const { result } = await queriesContract.simulate.querySwap(
+                [this.swaps as SingleSwap, DEFAULT_FUND_MANAGMENT],
+                { blockNumber: block },
+            );
+
+            amount =
+                this.swapKind === SwapKind.GivenIn
+                    ? TokenAmount.fromRawAmount(this.outputAmount.token, result)
+                    : TokenAmount.fromRawAmount(this.inputAmount.token, result);
+        }
+
+        return amount;
+    }
+
+    private convertNativeAddressToZero(address: Address): Address {
+        return address === NATIVE_ADDRESS ? ZERO_ADDRESS : address;
+    }
+
+    public queryCallData(): string {
+        let callData: string;
+        if (this.isBatchSwap) {
+            callData = encodeFunctionData({
+                abi: balancerQueriesAbi,
+                functionName: 'queryBatchSwap',
+                args: [
+                    this.swapKind,
+                    this.swaps as BatchSwapStep[],
+                    this.assets,
+                    DEFAULT_FUND_MANAGMENT,
+                ],
+            });
+        } else {
+            callData = encodeFunctionData({
+                abi: balancerQueriesAbi,
+                functionName: 'querySwap',
+                args: [this.swaps as SingleSwap, DEFAULT_FUND_MANAGMENT],
+            });
+        }
+        return callData;
+    }
+
+    public get priceImpact(): PriceImpactAmount {
+        const paths = this.pathsImmutable;
+
+        const pathsReverse = paths.map(
+            (path) =>
+                new PathWithAmount(
+                    this.chainId,
+                    [...path.tokens].reverse(),
+                    [...path.pools].reverse(),
+                    path.outputAmount.amount, // TODO - Is this correct scale?
```

Comment:
Yes, since you're using `TokenAmount.fromRawAmount()`, this seems to be the correct scaling 👍